cranelift/codegen/src/isa/x64/lower.isle

```diff
@@ -1457,8 +1457,9 @@
 ;; The flags must not have been clobbered by any other instruction between the
 ;; iadd_ifcout and this instruction, as verified by the CLIF validator; so we
 ;; can simply use the flags here.
-(rule (lower (trapif cc (iadd_ifcout _ _) tc))
-      (let ((_ Unit (emit_consumer (trap_if cc tc))))
+(rule (lower (trapif cc flags @ (iadd_ifcout _ _) tc))
+      (let ((producer ProducesFlags (ProducesFlags.AlreadyExistingFlags flags))
```

Comment:
I think this could work a little better if we did the following:

- The `AlreadyExistingFlags` enum arm has no args -- we don't actually hold onto the flags `Value`, it's just a sentinel representing "whatever's in the machine flags register";
- We write a helper in ISLE that does something like:
  ```lisp
  (decl flags_to_producesflags (Value) ProducesFlags)
  (rule (flags_to_producesflags flags)
    (mark_value_used flags)
    (ProducesFlags.AlreadyExistingFlags))
  ```

then we don't need the distinct case when emitting.

Reply:
And just to write down a thought I had and then dismissed: I had considered whether it made sense to make the above an auto-conversion, but I think that's too dangerous; if makes more sense to require an explicit action here since we're essentially asserting that `flags` is a flags value.

... or, perhaps we do:

```lisp
(rule (flags_to_producesflags (has_type $FLAGS flags))
  ...)
```

then we are actually asserting it's a flags value, and the verifier will ensure this corresponds to the current flags (there is only one live flags at a time), and so then this is safe and we can make it an auto-conversion. That removes all footguns from the lowerings I think.

Reply:
Would we need to select the second output of `flags` for that assertion to work, or am I missing something about how the outputs are accessed?

Reply:
I realized that I was forgetting that the dependency in `trapif` will explicitly be for the flags output of `iadd_ifcout`, so I was indeed missing something :)